### ISSUE TITLE
Limit in length for Variable Length Array Tables in FITS

### ISSUE THREAD

**Issue by vfioretti** [astropy/astropy#10812]
### Description
In writing a variable length table to a FITS file, it seems that there is a maximum allowed length for the arrays composing the column of the FITS (appearing to be 23171 despite different table sizes).
If the maximum length is exceeded, the data written in the FITS table is corrupted (no error messages appearing when running the script).
It would be very helpful if you could confirm the issue.
The possible workaround, i.e. creating a fixed length table and placing null values where the original array has no values, works but it doubles the FITS size which I am trying to avoid.

### Expected behavior
I made an example where a write to a FITS a table containing for each row an array long as the row number. 
So for example (nrows is the number of rows):
[[0], [0, 1], [0, 1, 2], [0, 1, 2, 3], ..., [0, 1, .., nrows-1]]

If nrows is less than 23171, the FITS is correctly written, and when I read the FITS file I get:
![test_23170](https://user-images.githubusercontent.com/4559399/95470485-25877e80-0981-11eb-9385-932cb500f1d7.png)

### Actual behavior

If nrows is higher than 23170 (nrows = 30000 in this case), I instead get corrupted rows where the index is higher than 23170:
![test_30000](https://user-images.githubusercontent.com/4559399/95470825-86af5200-0981-11eb-85ca-f42fc82a876f.png)

### Steps to Reproduce
Here the function to reproduce my example:
```
def write_fits(outname,nrows):
    
    from astropy.io import fits
    import numpy as np


    #=============================================================================================
    # a matrix with variable length array elements is created
    matrix = np.zeros(nrows, dtype=np.object_)
    xcol = np.ones(nrows)
    for i in range(0, nrows):
        xcol[i] = i
        matrix[i] = np.arange(0., np.float(i+1))
    
    #=============================================================================================
    
    col1 = fits.Column(name='x', format='J', unit='', array=xcol)
    col2 = fits.Column(name='MATRIX', format='PD('+str(nrows)+')', unit='', array=matrix)
    
    cols=fits.ColDefs([col1, col2])
    tbhdu = fits.BinTableHDU.from_columns(cols)
    tbhdu.name='MATRIX'
    tbhdu.writeto(outname,overwrite=True)

```

1. If matrix.fits is the name of the FITS file and nrows = 30000, to run the example just type
```
 write_fits("./matrix.fits", 30000)
```

### System Details

Python 3.8.5 (default, Sep  4 2020, 02:22:02) 
[Clang 10.0.0 ]
Numpy 1.19.1
astropy 4.0.1.post1
Scipy 1.5.0
Matplotlib 3.3.1

**Comment by saimn** [comment]
I think you're reaching the size limit of what can be indexed in the heap with PD, which is coded with 32-bit signed integer values. It seems to work correctly with QD, with uses 64-bit signed integer values. We should probably add a check for this to avoid corrupted files.

**Comment by vfioretti** [comment]
Yes! I confirm that with "Q" instead of "P" the issue is solved (below the image with the correct table).
I would also suggest to add this option in the wiki page:
https://docs.astropy.org/en/stable/io/fits/usage/unfamiliar.html
Thanks a lot for your quick and efficient feedback!

![test_30000_CORRECT](https://user-images.githubusercontent.com/4559399/95482769-c16bb700-098e-11eb-87a6-7626644d6e52.png)

**Comment by saimn** [comment]
Thanks for confirming that it works! Indeed Q should be mentioned in the docs. 
I will reopen the issue to keep in mind what needs to be done:
- add Q to the docs
- check if the heap size if bigger than what can be indexed with a 32 bit signed int.

**Comment by Cameron-Van-Eck** [comment]
I'd like to put in a 'bump' for this issue (or rather, for mentioning this behaviour in the documentation). @AlecThomson and I recently ran into this exactly this problem while developing code for storing radio spectra with varying numbers of channels (thus requiring variable length arrays). Our tables started generating unexpected behaviour when reading in very large (>2GB) tables.

I've confirmed that switching to the 'Q' column format appears to solve the problem. But Alec and I could have saved ourselves a few hours of debugging if this was mentioned in the documentation!

Thanks,
Cameron Van Eck
